### ISSUE TITLE
Bun TCP Sockets refuse to connect with hostname as "localhost"

### ISSUE THREAD

**Issue by PranjalAgni** [oven-sh/bun#7158]
### What version of Bun is running?

1.0.11+f7f6233ea

### What platform is your computer?

Darwin 23.1.0 x86_64 i386

### What steps can reproduce the bug?

I wrote a simple TCP server and client just to try the new API but client refuses(ECONNREFUSED) to connect when I set hostname as `localhost` in the server. Sharing the code that I am running:

server.ts
```ts
Bun.listen<SocketData>({
  hostname: "localhost",
  port: 4000,
  socket: {
    data(socket, data) {
      console.log("Ack: ", socket.data.sessionId);
      console.log("Client says: ", data.toString());
    }, // message received from client
    open(socket) {
      console.log("Client connected");
      socket.data = {
        sessionId: randomUUID(),
      };
      socket.write(socket.data.sessionId);
    }, // socket opened
    close(socket) {
      console.log("Client disconnected");
    }, // socket closed

  },
});
```

client.ts
```ts
Bun.connect({
  hostname: "localhost",
  port: 4000,
  socket: {
    data(socket, data) {
      console.log("Got data from server: ", data.toString());
    },
  },
});
```

Output:
<img width="515" alt="image" src="https://github.com/oven-sh/bun/assets/26196076/7d61be8b-c60c-4321-8d6f-6f58369147b0">

But as soon as I set the hostname as `127.0.0.1` on the server side it starts to work fine and socket connection/data exchange is getting made.

### What is the expected behavior?

Socket connection should be made irrespective of the hostname is set `localhost` or `127.0.0.1`

**Comment by Electroid** [comment]
Possibly related: #1425

**Comment by PranjalAgni** [comment]
I updated the docs for now, so the TCP sockets example works #7188

**Comment by Jarred-Sumner** [comment]
This was fixed by happy eyeballs #11206
